### PR TITLE
Fixed inconsistencies

### PR DESCRIPTION
**Pull request by naums**
Fixed inconsistencies; In the original file the struct name was used inconsistently, sometimes the real defined one, sometimes pyb_spi_obj. I also tried to extent the explenations and added comments to some code-lines to make them more accessible.

### CODE REVIEW

File: adding-module.rst [hunk 1]
```diff
@@ -62,15 +63,16 @@ structure. Open the ``mymodule.c`` file and put this code inside:
     #include "portmodules.h"
 
 
-    STATIC const mp_map_elem_t mymodule_globals_table[] = {
+    STATIC const mp_map_elem_t mymodule_globals_table[] = 
+    {
```
Comment (deshipu) [inline adding-module.rst]: We don't put braces on separate lines. Also, it would be nice to submit any code style changes in a separate pull request.

File: adding-module.rst [hunk 2]
```diff
@@ -10,6 +10,7 @@ will need to extend the existing firmware with your own C code, and this
 article is going to show you how to do it, using the ESP8266 port as the
 example.
 
+Note: you can use any naming-scheme you like and do not have to name your variables or data-type in any specific way.
```
Comment (deshipu) [inline adding-module.rst]: The fact that you can doesn't mean you should. Consistency with the existing code is important.

File: adding-module.rst [hunk 3]
```diff
@@ -62,15 +63,16 @@ structure. Open the ``mymodule.c`` file and put this code inside:
...
+    STATIC const mp_map_elem_t mymodule_globals_table[] = 
+    {
         { MP_OBJ_NEW_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_mymodule) },
     };
 
     STATIC MP_DEFINE_CONST_DICT(mp_module_mymodule_globals, mymodule_globals_table);
 
-    const mp_obj_module_t mp_module_mymodule = {
+    const mp_obj_module_t mp_module_mymodule = 
+    {
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 4]
```diff
@@ -106,8 +108,8 @@ add this code right after the includes:
 
     #include <stdio.h>
 
-
-    STATIC mp_obj_t mymodule_hello(void) {
+    STATIC mp_obj_t mymodule_hello(void) 
+    {
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 5]
```diff
@@ -116,16 +118,19 @@ add this code right after the includes:
 
 This creates a function object ``mymodule_hello_obj`` which takes no arguments,
 and when called, executes the C function ``mymodule_hello``. Also note, that
-our function has to return something -- so we return ``None``. Now we need to
+our function has to return something (as evey function returns an mp_obj_t-struct) -- so we return ``None``. Now we need to
```
Comment (deshipu) [inline adding-module.rst]: "every"
Please use a spellchecker before you submit a pull request.

It would be probably better to say "every Python function", because you can of course have functions of any type in C.

File: adding-module.rst [hunk 6]
```diff
@@ -116,16 +118,19 @@ add this code right after the includes:
...
 actually add that function object to our module:
 
 .. code-block:: c
 
-    STATIC const mp_map_elem_t mymodule_globals_table[] = {
+    STATIC const mp_map_elem_t mymodule_globals_table[] = 
+    {
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 7]
```diff
@@ -116,16 +118,19 @@ add this code right after the includes:
...
         { MP_OBJ_NEW_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_mymodule) },
         { MP_OBJ_NEW_QSTR(MP_QSTR_hello), (mp_obj_t)&mymodule_hello_obj },
     };
 
+Micropython uses the QSTR-macros to define constant strings. This is used to identify strings and store only unique ones for preserving memory (as it is very limited on the PyBoard-Hardware). Your port has a file ``qstrdefsport.h``. In our case add ``Q(hello)`` to the list (on a new line). This will define the string ``hello`` for Micropython. Failing to do so will result in a missing file on compilation. 
+
```
Comment (deshipu) [inline adding-module.rst]: This is done automatically during compilation, I believe.

File: adding-module.rst [hunk 8]
```diff
@@ -140,12 +145,14 @@ C function then needs to take an argument of type ``mp_obj_t``:
 
 .. code-block:: c
 
-    STATIC mp_obj_t mymodule_hello(mp_obj_t what) {
+    STATIC mp_obj_t mymodule_hello(mp_obj_t what) 
+    {
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 9]
```diff
@@ -140,12 +145,14 @@ C function then needs to take an argument of type ``mp_obj_t``:
...
         printf("Hello %s!\n", mp_obj_str_get_str(what));
         return mp_const_none;
     }
     STATIC MP_DEFINE_CONST_FUN_OBJ_1(mymodule_hello_obj, mymodule_hello);
 
+This function will use the C-function ``printf`` to output a string. The parameter ``what`` will be turned into a string by the ``mp_obj_str_get_str``-function (i.e. by Micropython). If you want to do that yourself, https://github.com/micropython/micropython/blob/master/py/obj.h#L472 (py/obj.h Line 472).
```
Comment (deshipu) [inline adding-module.rst]: Can you make this link a link?

Comment (deshipu) [reply adding-module.rst]: I am not sure what this last sentence is supposed to mean. Can you elaborate?

File: adding-module.rst [hunk 10]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
 Classes
 =======
 
-Let's try to add a class to our module. A class is similar to a module -- it's
-also a C struct with certain fields:
+A class is a C-struct with certain fields, quite similar to a module:
```
Comment (deshipu) [inline adding-module.rst:176]: How is this version better?

File: adding-module.rst [hunk 11]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
-
-    const mp_obj_type_t mymodule_hello_type = {
-        { &mp_type_type },
-        .name = MP_QSTR_Hello,
-        .print = mymodule_hello_print,
-        .make_new = mymodule_hello_make_new,
-        .locals_dict = (mp_obj_dict_t*)&mymodule_hello_locals_dict,
+    // creating the table of global members
+    STATIC const mp_rom_map_elem_t mymodule_hello_locals_dict_table[] = { };
+    STATIC MP_DEFINE_CONST_DICT(mymodule_hello_locals_dict, mymodule_hello_locals_dict_table);
```
Comment (deshipu) [inline adding-module.rst]: Please wrap at 80 characters.

File: adding-module.rst [hunk 12]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
+
+    // create the class-object itself
+    const mp_obj_type_t mymodule_helloObj_type = 
+    {
+        { &mp_type_type },                  // "inherit" the type "type"
+        .name = MP_QSTR_helloObj,           // give it a name
+        .print = mymodule_hello_print,      // give it a print-function
+        .make_new = mymodule_hello_make_new,    // give it a constructor
+        .locals_dict = (mp_obj_dict_t*)&mymodule_hello_locals_dict, // and the global members
```
Comment (deshipu) [inline adding-module.rst]: Please use consistent spacing with comments: two spaces, not some random indentations

File: adding-module.rst [hunk 13]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
-        .print = mymodule_hello_print,
-        .make_new = mymodule_hello_make_new,
-        .locals_dict = (mp_obj_dict_t*)&mymodule_hello_locals_dict,
+    // creating the table of global members
+    STATIC const mp_rom_map_elem_t mymodule_hello_locals_dict_table[] = { };
+    STATIC MP_DEFINE_CONST_DICT(mymodule_hello_locals_dict, mymodule_hello_locals_dict_table);
+
+    // create the class-object itself
+    const mp_obj_type_t mymodule_helloObj_type = 
+    {
```
Comment (deshipu) [inline adding-module.rst]: please don't put braces on their own lines

File: adding-module.rst [hunk 14]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
-
-    typedef struct _mymodule_hello_obj_t {
-        mp_obj_base_t base;
-        uint8_t hello_number;
+    
+    // this is the actual C-structure for our new object
+    typedef struct _mymodule_hello_obj_t 
+    {
+        mp_obj_base_t base;         // base represents some basic information, like type
+        uint8_t hello_number;       // a member created by us
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 15]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
     } mymodule_hello_obj_t;
 
 
-    mp_obj_t mymodule_hello_make_new(const mp_obj_type_t *type, size_t n_args,
-                                     size_t n_kw, const mp_obj_t *args) {
-        mp_arg_check_num(n_args, n_kw, 1, 1, true);
-        pyb_spi_obj_t *self = m_new_obj(mymodule_hello_obj_t);
+We define a C-struct which holds the class data and one additional field ``hello_number``. Next we need a function to print the object and a constructor:
```
Comment (deshipu) [inline adding-module.rst]: "C struct, which"

File: adding-module.rst [hunk 16]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
+    mp_obj_t mymodule_hello_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args)
+    {
+        // this checks the number of arguments (min 1, max 1);
+        // on error -> raise python exception
+        mp_arg_check_num(n_args, n_kw, 1, 1, true); 
+        // create a new object of our C-struct type
+        mymodule_hello_obj_t *self = m_new_obj(mymodule_hello_obj_t);
+        // give it a type
         self->base.type = &mymodule_hello_type;
+        // set the member number with the first argument of the constructor
```
Comment (deshipu) [inline adding-module.rst:230]: ditto about wrapping and spaces

File: adding-module.rst [hunk 17]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
 
-    STATIC void pyb_spi_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
-        pyb_spi_obj_t *self = MP_OBJ_TO_PTR(self_in);
-        mp_printf(print, "Hello(%u)", self->hello_number);
+    STATIC void mymodule_hello_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) 
+    {
+        // get a ptr to the C-struct of the object
+        mymodule_hello_obj_t *self = MP_OBJ_TO_PTR(self_in);
+        // print the number
+        printf ("Hello(%u)", self->hello_number);
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 18]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
+.. code-block:: c
 
-Of course you also need to add the class to the module's globals. Compile it
-and try creating and printing objects of our new class.
+    STATIC const mp_map_elem_t mymodule_globals_table[] = 
+    {
+        { MP_OBJ_NEW_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_mymodule) },
+        { MP_OBJ_NEW_QSTR(MP_QSTR_hello), (mp_obj_t)&mymodule_hello_obj },
+        { MP_OBJ_NEW_QSTR(MP_QSTR_helloObj), (mp_obj_t)&mymodule_helloObj_obj },
+    };
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 19]
```diff
@@ -159,58 +166,77 @@ modules already included in MicroPython. I will not be covering this in detail.
...
 
+Note that both the function ``mymodule_hello_obj`` added earlier to our module and the class ``mymodule_hello_obj`` are passed as ``mp_obj_t`` to the globals-table of the module.
```
Comment (deshipu) [inline adding-module.rst]: "globals table"

File: adding-module.rst [hunk 20]
```diff
@@ -221,21 +247,36 @@ is a pointer to the data struct:
 
 .. code-block:: c
 
-    STATIC mp_obj_t mymodule_hello_increment(mp_obj_t self_in) {
-        pyb_spi_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    STATIC mp_obj_t mymodule_hello_increment(mp_obj_t self_in) 
+    {
+        mymodule_hello_obj_t *self = MP_OBJ_TO_PTR(self_in);
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 21]
```diff
@@ -221,21 +247,36 @@ is a pointer to the data struct:
...
         self->hello_number += 1;
         return mp_const_none;
     }
-    MP_DEFINE_CONST_FUN_OBJ_1(mymodule_hello_increment_obj,
-                              mymodule_hello_increment);
+    MP_DEFINE_CONST_FUN_OBJ_1(mymodule_hello_increment_obj, mymodule_hello_increment);
```
Comment (deshipu) [inline adding-module.rst]: ditto

File: adding-module.rst [hunk 22]
```diff
@@ -221,21 +247,36 @@ is a pointer to the data struct:
...
 
 Also, don't forget to add them to the locals dict:
 
 .. code-block:: c
 
-    STATIC const mp_rom_map_elem_t mymodule_hello_locals_dict_table[] = {
-        { MP_ROM_QSTR(MP_QSTR_read), MP_ROM_PTR(&mymodule_hello_increment_obj) },
+    STATIC const mp_rom_map_elem_t mymodule_hello_locals_dict_table[] = 
+    {
+        { MP_ROM_QSTR(MP_QSTR_inc), MP_ROM_PTR(&mymodule_hello_increment_obj) },
```
Comment (deshipu) [inline adding-module.rst]: ditto

**Review by deshipu — CHANGES_REQUESTED** [review]
Thank you for the pull request, and sorry for taking so long to react -- I didn't see any notification from github, and I only now saw this.

A general advice for making pull requests (and commits too): try to fix one thing at a time. I have to reject this whole pull request, even though some parts of it really improve the documentation, because you also made a lot of unrelated changes that are not acceptable.

File: adding-module.rst (hunk 10 above)
Comment (naums) [reply adding-module.rst:176]: This is a change in the style of the text. I thought this would fit the rest of the style of the text better.

**Comment by naums** [comment]
I wanted to add, in the mean-time I wrote a blog post about adding modules to Micropython, where I tried to dig a bit deeper into the Micropython code, for example finding out what data types really are. This is probably not that exciting to you, but maybe you want to use some of that in this documentation-page. 
https://www.stefannaumann.de/en/2017/07/adding-a-module-to-micropython/

**Review by deshipu — APPROVED** [review]

**Comment by deshipu** [comment]
Thank you for that. I didn't have time to read your blog post yet, but certainly any way of growing the documentation would be good.
